### ISSUE TITLE
IBM DB2 SQL Error on rowlimit query

### ISSUE THREAD

**Issue by DavideVillaGEP** [Mgiepz/saiku-adhoc#98]
hi,
I'm testing the pentaho plugin.
I have no problem with the sample database and I made some report with it.
I'm trying to perform an adhoc report on a new JDBC SQL connection over a IBM DB2 server (AS400) and my DB server complain about a unrecognised keyword: TOP 100
I think the cause is the different handling of the row limit clause of the DB2 SQL dialect, so I try to perform the query with NONE as the rowlimit parameter but it seems this parameter remain set to 100.
is it possible to avoid passing the query limit to the generated SQL query?
this is my only blocker to setup a pentaho test server.
thanks.

**Comment by Mgiepz** [comment]
Regarding sql generation from mql to db2 the only info i could find is this
http://wiki.pentaho.com/display/PMOPEN/Limit

None = 100 definitely is a bug.  I just commited a fix so that it should be solved in the ci build

**Comment by DavideVillaGEP** [comment]
thanks for your fast reply,
I know that the AS400/DB2 dialect is problematic. instead of "TOP 100" I have to send "FETCH FIRST 100 ROWS ONLY"
but for my test is fine if I can disable the row limit.
I just installed the new ci build but the "TOP 100" clause is sent even if I set NONE in the row limit combo:
this is the query string:
SELECT DISTINCT TOP 100
BT_RAP990F_RAP990F.RAP_YEAR AS COL0
FROM 
RAP990F BT_RAP990F_RAP990F

and this is the pentaho.log:
Caused by: pt.webdetails.cda.dataaccess.QueryException: ReportDataFactoryException : The generated SQL-query did not execute successfully.; Parent exception: [SQL0104] Token TOP 100 is not valid: Valid tokens: + - AS <IDENTIFIER>.

```
at pt.webdetails.cda.dataaccess.PREDataAccess.performRawQuery(PREDataAccess.java:168)
at pt.webdetails.cda.dataaccess.SimpleDataAccess.queryDataSource(SimpleDataAccess.java:162)
... 67 more
```

Caused by: org.pentaho.reporting.engine.classic.core.ReportDataFactoryException: The generated SQL-query did not execute successfully.
    at org.pentaho.reporting.engine.classic.extensions.datasources.pmd.SimplePmdDataFactory.buildTableModel(SimplePmdDataFactory.java:412)
    at org.pentaho.reporting.engine.classic.extensions.datasources.pmd.SimplePmdDataFactory.queryData(SimplePmdDataFactory.java:491)
    at org.pentaho.reporting.engine.classic.extensions.datasources.pmd.PmdDataFactory.queryData(PmdDataFactory.java:99)
    at org.pentaho.reporting.engine.classic.core.CompoundDataFactory.queryStatic(CompoundDataFactory.java:135)
    at org.pentaho.reporting.engine.classic.core.CompoundDataFactory.queryData(CompoundDataFactory.java:85)
    at org.pentaho.reporting.engine.classic.core.cache.CachingDataFactory.queryInternal(CachingDataFactory.java:424)
    at org.pentaho.reporting.engine.classic.core.cache.CachingDataFactory.queryData(CachingDataFactory.java:325)
    at pt.webdetails.cda.dataaccess.PREDataAccess.performRawQuery(PREDataAccess.java:150)
    ... 68 more
2012-02-16 16:20:57,624 ERROR [org.saiku.adhoc.rest.QueryResource] Cannot generate report (A26FBCF1-4429-23C8-8E26-BAF9BE53EDE5)
org.pentaho.reporting.engine.classic.core.ReportDataFactoryException: Failed to send request
    at org.pentaho.reporting.engine.classic.extensions.datasources.cda.HttpQueryBackend.fetchData(HttpQueryBackend.java:108)
    at org.pentaho.reporting.engine.classic.extensions.datasources.cda.CdaQueryBackend.queryData(CdaQueryBackend.java:184)
    at org.pentaho.reporting.engine.classic.extensions.datasources.cda.CdaDataFactory.queryData(CdaDataFactory.java:144)
    at org.pentaho.reporting.engine.classic.core.CompoundDataFactory.queryFreeForm(CompoundDataFactory.java:114)
    at org.pentaho.reporting.engine.classic.core.cache.CachingDataFactory.queryInternal(CachingDataFactory.java:418)
    at org.pentaho.reporting.engine.classic.core.cache.CachingDataFactory.queryFreeForm(CachingDataFactory.java:243)
    at org.pentaho.reporting.engine.classic.core.CompoundDataFactory.queryFreeForm(CompoundDataFactory.java:109)
    at org.pentaho.reporting.engine.classic.core.CompoundDataFactory.queryData(CompoundDataFactory.java:90)
    at org.pentaho.reporting.engine.classic.core.states.datarow.ReportDataRow.createDataRow(ReportDataRow.java:97)
    at org.pentaho.reporting.engine.classic.core.states.datarow.DefaultFlowController.performQuery(DefaultFlowController.java:188)
    at org.saiku.adhoc.service.report.ReportGeneratorService.processReport(ReportGeneratorService.java:272)
    at org.saiku.adhoc.service.report.ReportGeneratorService.renderReportHtml(ReportGeneratorService.java:156)
    at org.saiku.adhoc.rest.QueryResource.generateReport(QueryResource.java:261)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:601)

**Comment by Mgiepz** [comment]
could you please check in pentaho-solutions\system\saiku-adhoc\temp\ the latest temporary cda file that is created after you execute your query with limit=none? what does it say about the limit? if there is not a -1 something is wrong.

**Comment by DavideVillaGEP** [comment]
I made further investigations:
I checked the generated cda and the first time it generate the limit 100 even if I leave none in the toolbar combo.
then I try to set 1000 and the cda changed accordingly, then I set back to NONE and finally the cda changed to -1.
with that setting the query run well and my DB return the rows correctly.
so it seems the problem is only for the first time the combo says NONE but the cda is generated with 100.
for my tests the workaround is switch the combo to 1000 and then again to NONE so the generated query works like a charm.
thanks
